Speedometer 3: WebContent spends a lot of time in RemoteImageBufferProxy::ensureBackendCreated.

mattwoodrow opened the pull request:
#### e0a42565e3a2879ec1e7c1bc87937dc2d8b779eb
<pre>
Speedometer 3: WebContent spends a lot of time in RemoteImageBufferProxy::ensureBackendCreated.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266728">https://bugs.webkit.org/show_bug.cgi?id=266728</a>
&lt;<a href="https://rdar.apple.com/117823468">rdar://117823468</a>&gt;

Reviewed by Kimmo Kinnunen.

Synchronously waiting on the ImageBufferBackendHandle for an ImageBufferSet can block the main thread while
trying to serialize a transaction.

This splits the response to prepareImageBufferSetsForDisplay into two parts, one that is only present when needed
as a sync reponse, and another as a separate message returned asynchronously.

The new response message &apos;DidPrepareForDisplay&apos; stores the resulting buffer backend handles on the
RemoteImageBufferSetProxy and notifies a Condition, so that they can be waited on from a different thread.

RemoteLayerTree transaction serialization on the main thread now just serializes an identifier for the image
buffer set and doesn&apos;t block.

The transaction flusher thread waits for any incoming backend handles (as well as the existing wait on drawing
flushing), and builds a hashamp of image buffer set identifier to response data and appends this on to the
remote layer tree transaction.

RemoteLayerTreeDrawingAreaProxy uses the trailing hashmap to copy across any needed backend handles from the
response hashmap into the appropriate RemoteLayerBackingStoreProperties.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplaySync):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
(WebKit::RemoteRenderingBackend::identifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.h:
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::bufferSetIdentifier const):
(WebKit::RemoteLayerBackingStoreProperties::bufferSetIdentifier):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::setBackendHandle):
(WebKit::RemoteLayerBackingStore::takePendingFlushers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::didFlushLayers):
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeEncoded): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlusher):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlushers): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createFlusher):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::bufferSetIdentifier const):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createFlushers): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxyFlushFence::waitFor):
(WebKit::RemoteImageBufferSetProxyFlushFence::setHandles):
(WebKit::RemoteImageBufferSetProxyFlushFence::takeHandles):
(WebKit::RemoteImageBufferSetProxy::didPrepareForDisplay):
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
(WebKit::RemoteImageBufferSetProxy::remoteBufferSetWasDestroyed):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.messages.in: Copied from Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackendProxy::didPrepareForDisplay):
(WebKit::RemoteRenderingBackendProxy::dispatchMessage):
(WebKit::RemoteRenderingBackendProxy::ensurePrepareCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::BackingStoreFlusher::flush):

Canonical link: <a href="https://commits.webkit.org/273323@main">https://commits.webkit.org/273323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fb2ee8be1d97646cf57280f5f510ab69ef7c537

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35051 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37146 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37810 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31658 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16342 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11035 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11828 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10366 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/35313 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10426 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39058 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31891 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31694 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36439 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10531 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8463 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34441 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12322 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31010 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8037 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11061 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11386 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -515,6 +514,24 @@ bool RemoteRenderingBackendProxy::isCached(const ImageBuffer& imageBuffer) const
 
+void RemoteRenderingBackendProxy::imageBufferSetDidPrepareForDisplay(WebKit::RemoteImageBufferSetIdentifier bufferSet, WebKit::ImageBufferSetPrepareBufferForDisplayOutputData outputData)
+{
+    m_background->putImageBufferSetPrepareForDisplayOutputData(bufferSet, WTFMove(outputData));
```
This is a message to the buffer set: RemoteImageBuffersSetProxy::DidPrepareForDisplay.
Compare to same function RemoteImaegBufferProxy::DidCreateBackend is to RemoteImageBufferProxy.

This could be handled in the same way: forwarding the message to  RemoteImageBuffersSetProxy.
This would remove the confusing class RemoteRenderingBackendProxyBackground.

mattwoodrow replied:
That's fair. The complication is that we need to keep those `RemoteImageBufferSetProxy` objects alive while we wait, since the main-thread will be free to run other code, which might remove layers.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -515,6 +514,24 @@ bool RemoteRenderingBackendProxy::isCached(const ImageBuffer& imageBuffer) const
+
+ImageBufferSetPrepareBufferForDisplayOutputData RemoteRenderingBackendProxyBackground::takeImageBufferSetPrepareForDisplayOutputData(RemoteImageBufferSetIdentifier bufferSet)
+{
+    auto handles = m_imageBufferSetsOutputData.take({ { bufferSet, 0 }, 0 }, IPC::Timeout::infinity());
```
the threadsafeobjectheap is used here "incorrectly", not for purpose of what it is created. It's a lock+cv substitute for RemoteImageBuffersSetProxy.

Also: we should not wait for infinity. It may well be that the message never arrives.

instead of object heap, see the comment in the fence

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
```diff
@@ -83,20 +85,42 @@ class RemoteImageBufferSetProxyFlushFence : public ThreadSafeRefCounted<RemoteIm
 public:
-    RemoteImageBufferSetProxyFlusher(Ref<RemoteImageBufferSetProxyFlushFence> flushState)
-        : m_flushState(WTFMove(flushState))
+    RemoteImageBufferSetProxyFlusher(RemoteImageBufferSetIdentifier identifier, Ref<RemoteImageBufferSetProxyFlushFence> flushState, Ref<RemoteRenderingBackendProxyBackground> background, unsigned generation)
```
This flush fence does now two things:
- wait until flush has completed
- wait until the handles have arrived.

The currently matching implementation for RemoteImageBufferSetProxy::didPrepareForDisplay would be:
```
class RemoteImageBufferSetProxyFlushFence ... { 
    bool waitFor(Seconds timeout)
    {
        Locker locker { m_lock };
        if (!m_signaled) {
            m_signaled = m_event.waitFor(timeout);
            if (m_signaled)
                 tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 0u);
         }
        if (m_signaled &&!m_backendHandles)
           m_condition.waitUntil(...);
        return m_signaled && m_backendHandles;
    }
    void setBackendHandles(..) {
        Locker locker { m_lock };
        m_backendHandles = ...;
        m_condition.signalAll(); 
      )
};

void RemoteImageBufferSetProxy::didPrepareForDisplay(outputdata) (
    if (!m_pendingFlush)
       return;
    m_pendingFlush->setBackendHandles(outputData);
}
```


E.g. store the backend handles that need to be waited in the wait object shared state.

however: the code doesn't handle the case where wait for the handles or flush times out, the code progresses and flushes again, and then two flushes complete: the old timed out one and the new correct one..

mattwoodrow replied:
That last bit could indeed be a problem. I might need to add an identifier to the prepareForDisplay call, returning into the didPrepareForDisplay response, to identify if it's the one we're currently waiting for.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
```diff
@@ -167,6 +195,30 @@ void RemoteImageBufferSetProxy::clearVolatilityUntilAfter(MarkSurfacesAsVolatile
+
+        handle.bufferHandle = WTFMove(outputData.backendHandle);
+
+        auto createBufferAndBackendInfo = [&](const  std::optional<WebCore::RenderingResourceIdentifier>& bufferIdentifier) {
```
two spaces..

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxyBackground.h:
```diff
@@ -0,0 +1,58 @@
+/*
```
This should be removed?

mattwoodrow committed:
Speedometer 3: WebContent spends a lot of time in RemoteImageBufferProxy::ensureBackendCreated.
https://bugs.webkit.org/show_bug.cgi?id=266728
<rdar://117823468>

Reviewed by Kimmo Kinnunen.

Synchronously waiting on the ImageBufferBackendHandle for an ImageBufferSet can block the main thread while
trying to serialize a transaction.

This splits the response to prepareImageBufferSetsForDisplay into two parts, one that is only present when needed
as a sync reponse, and another as a separate message returned asynchronously.

The new response message 'DidPrepareForDisplay' stores the resulting buffer backend handles on the
RemoteImageBufferSetProxy and notifies a Condition, so that they can be waited on from a different thread.

RemoteLayerTree transaction serialization on the main thread now just serializes an identifier for the image
buffer set and doesn't block.

The transaction flusher thread waits for any incoming backend handles (as well as the existing wait on drawing
flushing), and builds a hashamp of image buffer set identifier to response data and appends this on to the
remote layer tree transaction.

RemoteLayerTreeDrawingAreaProxy uses the trailing hashmap to copy across any needed backend handles from the
response hashmap into the appropriate RemoteLayerBackingStoreProperties.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackend::prepareImageBufferSetsForDisplaySync):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
(WebKit::RemoteRenderingBackend::identifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(class_template_headers):
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.h:
* Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::bufferSetIdentifier const):
(WebKit::RemoteLayerBackingStoreProperties::bufferSetIdentifier):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::setBackendHandle):
(WebKit::RemoteLayerBackingStore::takePendingFlushers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::didFlushLayers):
(WebKit::RemoteLayerBackingStoreCollection::backingStoreWillBeEncoded): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlusher):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::createFlushers): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createFlusher):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::bufferSetIdentifier const):
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::createFlushers): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.cpp:
(WebKit::RemoteImageBufferSetProxyFlushFence::waitFor):
(WebKit::RemoteImageBufferSetProxyFlushFence::setHandles):
(WebKit::RemoteImageBufferSetProxyFlushFence::takeHandles):
(WebKit::RemoteImageBufferSetProxy::didPrepareForDisplay):
(WebKit::RemoteImageBufferSetProxy::flushFrontBufferAsync):
(WebKit::RemoteImageBufferSetProxy::remoteBufferSetWasDestroyed):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferSetProxy.messages.in: Copied from Source/WebKit/Shared/RemoteLayerTree/BufferAndBackendInfo.serialization.in.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::prepareImageBufferSetsForDisplay):
(WebKit::RemoteRenderingBackendProxy::didPrepareForDisplay):
(WebKit::RemoteRenderingBackendProxy::dispatchMessage):
(WebKit::RemoteRenderingBackendProxy::ensurePrepareCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::BackingStoreFlusher::flush):

Canonical link: https://commits.webkit.org/273323@main
